neqo-common/src/codec.rs

```diff
@@ -92,33 +82,43 @@ impl<'a> Decoder<'a> {
         Some(res)
     }
 
-    /// Decodes an unsigned integer of length 1..=8.
-    ///
-    /// # Panics
-    ///
-    /// This panics if `n` is not in the range `1..=8`.
-    pub fn decode_uint(&mut self, n: usize) -> Option<u64> {
-        assert!(n > 0 && n <= 8);
+    #[inline]
+    pub(crate) fn decode_n(&mut self, n: usize) -> Option<u64> {
+        debug_assert!(n > 0 && n <= 8);
         if self.remaining() < n {
             return None;
         }
-        let mut v = 0_u64;
-        for i in 0..n {
-            let b = self.buf[self.offset + i];
-            v = v << 8 | u64::from(b);
-        }
-        self.offset += n;
-        Some(v)
+        Some(if n == 1 {
```

Comment:
This looks pretty bad, but it's the reason for the largest part of the performance wins.  The other change is a 5% improvement and maybe even a regression in some cases, but this if statement leads to huge gains.

Reply:
Oh, I would have expected replacing the `for` loop with the `copy_from_slice` to have the highest impact.